main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+    *
+    * We will often write a substitution as `[x -> t1, y -> t2]` and so forth. We write that `x -> t1` is a binding.
+    *
+    * Note: constants and variables are a separate syntactic category. A substitution will never replace any constants.
+    */
+  case class BoolSubstitution(m: Map[Int, Term]) {
+
+    /**
+      * Applies `this` substitution to the given term `t`.
+      *
+      * We must use the smart constructors from [[Term]] to ensure that the constructed term is normalized.
+      */
+    def apply(t: Term): Term = t match {
+      case Term.True => Term.True
+      case Term.False => Term.False
+      case Term.Cst(c) => Term.Cst(c)
+
+      case Term.Var(x) => m.get(x) match {
+        case None => Term.Var(x) // Case 1: The substitution has a binding for `x`. Return the bound term.
+        case Some(t0) => t0 // Case 2: The substitution has no binding for `x`. Return the original term.
+      }
+
+      case Term.Not(t0) => Term.mkNot(apply(t0))
+
+      case Term.And(csts, vars, rest) =>
+        // A conjunction is a sequence of: constants, variables, and terms. We know that the constants are unchanged by
+        // the substitution. We know that some variables may become constants, variables, or terms. Since we do not want
+        // to duplicate functionality from [[Term.mkAnd]], we simply apply the substitution to the variables and terms,
+        // and put everything in one list. We then let [[Term.mkAnd]] reclassify all the sub-terms.
+        val ts = csts.toList ++ vars.toList.map(apply) ++ rest.map(apply)
+        Term.mkAnd(ts)
+
+      case Term.Or(ts) => Term.mkOr(ts.map(apply))
+    }
+
+    /**
+      * Applies `this` substitution to the given equation `e`.
+      *
+      * Intuitively, applies `this` to the lhs and rhs of `e` and reconstructs the equation.
+      *
+      * Applying the substitution and reconstructing the equation may "flip" the lhs and rhs. For example:
+      *
+      * If `s = [x -> y]` and `e = true ~ x and y` then `s(e) = y ~ true` which has flipped lhs and rhs.
+      */
+    def apply(e: Equation): Equation = e match {
+      case Equation(t1, t2, loc) => Equation.mk(apply(t1), apply(t2), loc)
+    }
+
+    /**
+      * Applies `this` substitution to the given list of equations `l`.
+      */
+    def apply(l: List[Equation]): List[Equation] = l.map(apply)
+
+    /**
+      * Returns the number of bindings in `this` substitution.
+      */
+    def numberOfBindings: Int = m.size
+
+    /**
+      * Returns the size of `this` substitution.
+      *
+      * The size of a substitution is the sum of the sizes of all terms in its co-domain.
+      */
+    def size: Int = m.values.map(_.size).sum
+
+    /**
+      * Extends `this` substitution with a new binding from the variable `x` to the term `t`.
+      *
+      * If the variable `x` already occurs in `this` substitution then it must be bound to the same term as `t`.
```

Comment:
When we use "must", it usually means that a violation results in an ICE. Here it seems that violation might sometimes be expected, so we should just say a Conflict exception is thrown

Reply:
Good point, comment fixed.